fabtests/functional/stream_msg.c

```
@@ -0,0 +1,273 @@
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <getopt.h>
+#include "shared.h"
+#include <arpa/inet.h>
+#include <netdb.h>
+#include <string.h>
+#include <rdma/fi_cm.h>
+
+const char *msg = "hello stream!";
+
+
+int send_stream(struct fid_ep *ep, const char *msg, size_t msg_len)
+{
+	int offset, ret;
+
+	for (offset = 0; offset < msg_len; ) {
+		ret = fi_send(ep, (msg + offset), (msg_len - offset), NULL, 0, NULL);
+		if (ret < 0 && ret != -FI_EAGAIN) {
+			fprintf(stderr, "%s error %s\n", __func__, fi_strerror(-ret));
+			return ret;
+		}
+
+		if (ret > 0)
+			offset += ret;
+	}
+
+	return offset;
+}
+
+int recv_stream(struct fid_ep *ep, char *msg, size_t msg_len)
+{
+	int offset, ret;
+
+	for (offset = 0; offset < msg_len; ) {
+		ret = fi_recv(ep, (msg + offset), (msg_len - offset), NULL, 0, NULL);
+		if (ret < 0 && ret != -FI_EAGAIN) {
+			fprintf(stderr, "%s error %s\n", __func__, fi_strerror(-ret));
+			return ret;
+		}
+		if (ret > 0)
+			offset += ret;
+	}
+
+	return offset;
+}
+
+int static send_greeting(struct fid_ep *ep)
+{
+	const size_t msg_len = strlen(msg);
+	char buffer[msg_len];
+	int ret;
+
+	ret = send_stream(ep, msg, msg_len);
+	if (ret < 0)
+		return ret;
+
+	ret = recv_stream(ep, buffer, msg_len);
+	if (ret < 0)
+		return ret;
+
+	if (strncmp(buffer, msg, msg_len) != 0)
+		printf("error recv: %s\n", buffer);
```

If buffer != msg, would it be better to return an error code?

---

updated - thanks